SYCL: Refactor ggml_sycl_compute_forward

qnixsynapse opened the pull request:
I originally thought I wouldn't do this, but the original code is quite restrictive imo. It assumes that all operations need only src0, src1 and dst tensors, which isn't the case for every OP(also, why are we passing them separately?). This change will allow easy addition of tensor operations to the backend as the `dst` tensor itself contains everything needed to perform the operation. I haven't changed the kernels except for a few function signature changes, as that's outside the scope of this refactoring. Those can be refactored later if needed.

qnixsynapse committed:
SYCL: refactor ggml_sycl_compute_forward

qnixsynapse committed:
SYCL: add back GGML_USED(dst) to ggml_sycl_cpy

qnixsynapse committed:
SYCL: add function name to noop debug

Alcpz reviewed (approved):
Code looks simpler now. Thanks for the change. 

Edit: Will wait for other to give a review.

qnixsynapse commented:
@Alcpz Thanks for your review.
Tagging @Rbiessy @NeoZhangJianyu @airMeng manually to get their thoughts on this.

qnixsynapse committed:
SYCL: Some device info print refactoring and add details of XMX availability

qnixsynapse commented:
I ended up refactoring the device info printing and added if a device has XMX or tensor cores. I used sycl::aspect::ext_intel_matrix to determine whether an Intel GPU has XMX, but I'm not sure how to determine this for Nvidia and AMD GPUs. If you know how to determine this for Nvidia and AMD GPUs, please let me know.
Currently device info with this change displays like this:
![Screenshot From 2025-01-09 12-06-58](https://github.com/user-attachments/assets/861916c1-c391-4b35-988d-a18a393332d9)

Added this to PR, since I want to focus on optimizing the kernels next.

Alcpz commented:
Thanks for the further contribution. However, please refrain from adding more functionality once you've got approvals. This one is harmless, but it's a new change nonetheless.

Rbiessy commented on ggml/src/ggml-sycl/ggml-sycl.cpp:
```diff
@@ -54,18 +54,12 @@ static ggml_sycl_device_info ggml_sycl_init() {
-#endif
-    GGML_LOG_INFO("%s: found %d %s devices:\n", __func__, info.device_count, GGML_SYCL_NAME);
-
+/* This is a bit misleading;  reserved for later */
+// #if defined(SYCL_USE_XMX)
+//     GGML_LOG_INFO("%s: SYCL_USE_XMX: yes\n", __func__);
+// #else
+//     GGML_LOG_INFO("%s: SYCL_USE_XMX: no\n", __func__);
+// #endif
```
As far as I can tell the project never queries for `sycl::aspect::ext_intel_matrix` but relies on the macro `SYCL_USE_XMX` to change some tiles parameters. As this [TODO](https://github.com/ggerganov/llama.cpp/blob/master/ggml/src/ggml-sycl/common.hpp#L68) says it is not a great name for it but I think it makes sense to log whether `SYCL_USE_XMX` is set or not.
I think printing whether the device has `sycl::aspect::ext_intel_matrix` is more misleading.

qnixsynapse replied:
`SYCL_USE_XMX` is only set when `GGML_SYCL_FORCE_MMQ` is not defined:

https://github.com/ggerganov/llama.cpp/blob/f8feb4b01af374ad2fce302fd5790529c615710b/ggml/src/ggml-sycl/common.hpp#L67-L71

The comment mentions that it is not being used for SYCL currently. It also seems like it is always true since mmq kernels are disabled/faulty.

`sycl::aspect::ext_intel_matrix` checks whether hardware supports XMX or not. I am not sure how that is misleading. We can re-enable these `SYCL_USE_XMX` after proper implementation.

Rbiessy replied:
Thanks for the explanations. Maybe it's not that useful to log for `SYCL_USE_XMX` as long as it is tied to `GGML_SYCL_FORCE_MMQ` which is already logged.

I wasn't sure what was the point for querying `sycl::aspect::ext_intel_matrix` but if some dependencies rely on it why not.

qnixsynapse replied:
Currently, `sycl::aspect::ext_intel_matrix` is only used for displaying if a GPU has XMX or not in `ggml_backend_sycl_print_sycl_devices` function replacing `SYCL_USE_XMX: yes/No` logs.

I want to expand this to other GPUs too but I am having hard time finding proper documentation.
In the future, I was thinking to cache it and make it useful for writing kernels with joint_matrix extensions for such devices.

Let's see what Intel team has to say. This is not important for now and can be removed.

Rbiessy reviewed (commented):
I have no concern with the original changes but it should probably be reviewed by the Intel team as well.

airMeng reviewed (approved):
Long time no see. Thank you!

airMeng commented:
Could you rebase and trigger the CI again?

qnixsynapse commented:
> Long time no see. 
@airMeng Likewise... :)
